### DISCUSSION TITLE
How to make format on save work with prettier?

### DISCUSSION THREAD

**Discussion by trozler** [PaulRBerg/hardhat-template#46] [Q&A]
Hi thanks for the template!

I'm struggling to getting format on save working for `*.sol` files 😓.
I have created a hack by putting the following in my workspace settings (format on save is on as well), and things are working.
```
{
  "editor.defaultFormatter": "esbenp.prettier-vscode"
}

```

Am I missing something very obvious? Or is the project not intended to format solidity contract on save?

**Comment by PaulRBerg** [comment]
Hi @trozler, glad you like the template.

Perhaps you're missing the following setting?

```json
{
  "editor.formatOnSave": true
}
```

**Reply by PaulRBerg** [reply]
Hey @trozler, did my solution work?

**Reply by trozler** [reply]
Hi thanks for the reply.  Unfortunately it did not. 
Looking at the prettier logs:
```
["INFO" - 13:27:53] Extension Name: esbenp.prettier-vscode.
["INFO" - 13:27:53] Extension Version: 6.4.0.

```

The extension seems to load but is never activated. 

Steps to replicate (only tested on my machine granted):
```
gh repo clone paulrberg/solidity-template
cd template
yarn install
```

Then open editor and try to format on save in a `*.sol` file.

What is interesting is the formatting on save works for `*.ts` files. 

```
INFO" - 13:33:14] Prettier Options:
{
  "filepath": "/Users/tonyrosler/Desktop/dev/testrepos/testTemplate/solidity-template/types/index.ts",
  "parser": "typescript",
  "useTabs": false,
  "tabWidth": 2,
  "endOfLine": "auto",
  "arrowParens": "avoid",
  "bracketSpacing": true,
  "printWidth": 120,
  "singleQuote": false,
  "trailingComma": "all"
}
["INFO" - 13:33:14] Formatting completed in 11.966502ms.
```

**Reply by PaulRBerg** [reply]
Tagging @fvictorio in case he might know what's going on here .. He's the creator of the prettier plugin for Solidity.

In the meantime, did you check that you're using the local version of Prettier?

```json
{
  "prettier.resolveGlobalModules": false,
}
```

**Reply by fvictorio** [reply]
Hmm, this depends a lot on the editor/IDE used. If you are using VSCode and the Solidity extension from Juan Blanco, then check its options, because it might be using the version of prettier-solidity that comes with it instead of the one you have installed. (I'm just guessing here, I don't use VSCode)

**Reply by trozler** [reply]
@fvictorio and @paulrberg thanks for the reply guys. 
So after some digging in my global settings I found the following lines:
```
  "[typescript]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[typescriptreact]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
```

I don't recall why I put them there, but there is some reason. So I simply added the following to my global settings:
```
  "[solidity]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
```

And everything works well !

NOTE:
@paulrberg I did try your suggestion about `{ "prettier.resolveGlobalModules": false,}`, which in my case did not work.
